### PR TITLE
[SNOW-647398] show details when printing a snowpark session

### PR DESCRIPTION
**Pull request by sfc-gh-stan**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes [SNOW-647398](https://snowflakecomputing.atlassian.net/browse/SNOW-647398)

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

It implements a `__repr__` method for Snowpark.Session so `print(session)` will display some helpful information about the session.
Before: `<snowflake.snowpark.session.Session object at 0x11c681700>`
After: `<snowflake.snowpark.session.Session: role='ROLE', database='DB', schema='SCHEMA', warehouse='WH'>`

### CODE REVIEW

File: src/snowflake/snowpark/session.py [hunk 1]
```diff
@@ -265,6 +265,9 @@ def __enter__(self):
     def __exit__(self, exc_type, exc_val, exc_tb):
         self.close()
 
+    def __repr__(self):
+        return f"<{self.__class__.__module__}.{self.__class__.__name__}: role={repr(self._conn._conn.role)}, database={repr(self._conn._conn.database)}, schema={repr(self._conn._conn.schema)}, warehouse={repr(self._conn._conn.warehouse)}>"
```
Comment (sfc-gh-kbregula) [inline src/snowflake/snowpark/session.py:269]: Should we also add an account?

Comment (sfc-gh-stan) [reply src/snowflake/snowpark/session.py:269]: Makes sense to me, I will add account in front of role.

File: src/snowflake/snowpark/session.py [hunk 2]
```diff
@@ -265,6 +265,9 @@ def __enter__(self):
     def __exit__(self, exc_type, exc_val, exc_tb):
         self.close()
 
+    def __repr__(self):
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/session.py:268]: I will prefer to implement `__str__` instead of `__repr__`. If `__repr__` gets implemented, we have to make sure `eval(repr(x)) == x`. https://stackoverflow.com/questions/1436703/what-is-the-difference-between-str-and-repr

File: src/snowflake/snowpark/session.py [hunk 3]
```diff
@@ -265,6 +265,9 @@ def __enter__(self):
...
+        return f"<{self.__class__.__module__}.{self.__class__.__name__}: account={repr(self._conn._conn.account)}, role={repr(self._conn._conn.role)}, database={repr(self._conn._conn.database)}, schema={repr(self._conn._conn.schema)}, warehouse={repr(self._conn._conn.warehouse)}>"
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/session.py:269]: 1. `self._conn._conn.role` and other similar properties can be None. You should use `get_current_*` methods in session class
2. I don't think you need `repr(x)` then.
3. Add tests when these properties are None (`get_current_*` methods  can also return None).

Comment (sfc-gh-kbregula) [reply src/snowflake/snowpark/session.py:269]: When can roles and other properties be empty? It seems to me that the `snowflake-connector-python` maintains the contents of these variables while logging in and executing SQL queries.
https://github.com/snowflakedb/snowflake-connector-python/blob/d0efd0f8333db210a3c482b6f631fe4ba01eabdc/src/snowflake/connector/connection.py#L1013-L1026

Comment (sfc-gh-stan) [reply src/snowflake/snowpark/session.py:269]: 1+2. I will use change the implementation to use `get_current_*` for better consistency. 
3. I agree with @sfc-gh-kbregula that for our use case, `get_current_*` will always short circuit at `getattr(self._conn, param) ` and return default values that the Python connector connection store.

Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/session.py:269]: When we create a session in Snowflake, database/schema/warehouse is optional. If we don't set them in connection parameters of python connector, then these properties will be empty.

Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/session.py:269]: nope, it's possible that the user doesn't have a default database/schema/warehouse. (For example, if you log into snowflake at the first time and don't create them before, how can we have defaults?)

Comment (sfc-gh-stan) [reply src/snowflake/snowpark/session.py:269]: I see, that makes sense 👍

**Review by sfc-gh-jdu — APPROVED** [review]

File: src/snowflake/snowpark/session.py [hunk 4]
```diff
@@ -265,6 +265,9 @@ def __enter__(self):
     def __exit__(self, exc_type, exc_val, exc_tb):
         self.close()
 
+    def __str__(self):
+        return f"<{self.__class__.__module__}.{self.__class__.__name__}: account={self.get_current_account()}, role={self.get_current_role()}, database={self.get_current_database()}, schema={self.get_current_schema()}, warehouse={self.get_current_warehouse()}>"
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/session.py:269]: Just curious if we could line break this in code? e.g.

```
return (f"<{self.__class__.__module__}.{self.__class__.__name__}: "
        f"account={self.get_current_account()}, "
        ...)

```

**Review by sfc-gh-sfan — APPROVED** [review]

File: src/snowflake/snowpark/session.py (hunk 4 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/session.py:269]: True, that looks a lot better, I wonder why flake8 doesn't catch this.
